src/stm32l0.c

```diff
@@ -0,0 +1,1076 @@
+      return -1;
+    if (sr & STM32Lx_NVM_SR_BSY)
+      continue;
+    if ((sr & STM32Lx_NVM_SR_ERR_M) || !(sr & STM32Lx_NVM_SR_EOP))
+      return -1;
+    break;
+  }
+
+  return 0;
+}
+
+
+/** Write one option word.  The address is the physical address of the
+    word and the value is a complete word value.  The caller is
+    responsible for making sure that the value satisfies the proper
+    format where the upper 16 bits are the 1s complement of the lower
+    16 bits.  The funtion returns when the operation is complete.
+    The return value is true if the write succeeded. */
+static bool stm32lx_option_write (target *t, uint32_t address, uint32_t value)
+{
+  ADIv5_AP_t*    ap  = adiv5_target_ap(t);
+  const uint32_t nvm = stm32lx_nvm_phys (t);
+
+  /* Erase and program option in one go. */
+  adiv5_ap_mem_write (ap, STM32Lx_NVM_PECR(nvm), STM32Lx_NVM_PECR_FIX);
+  adiv5_ap_mem_write (ap, address, value);
+
+  uint32_t sr;
+  do {
+    sr = adiv5_ap_mem_read (ap, STM32Lx_NVM_SR(nvm));
+  } while (sr & STM32Lx_NVM_SR_BSY);
+
+  return !(sr & STM32Lx_NVM_SR_ERR_M);
+}
+
+
+/** Write one eeprom value.  This version is more flexible than that
+    bulk version used for writing data from the executable file.  The
+    address is the physical address of the word and the value is a
+    complete word value.  The funtion returns when the operation is
+    complete.  The return value is true if the write succeeded.
+    FWIW, byte writing isn't supported because the adiv5 layer
+    doesn't support byte-level operations. */
+static bool stm32lx_eeprom_write (target *t, uint32_t address,
+                                  size_t cb, uint32_t value)
+{
+  ADIv5_AP_t*    ap         = adiv5_target_ap(t);
+  const uint32_t nvm        = stm32lx_nvm_phys (t);
+  const bool     is_stm32l1 = stm32lx_is_stm32l1 (t);
+
+  /* Clear errors. */
+  adiv5_ap_mem_write (ap, STM32Lx_NVM_SR (nvm), STM32Lx_NVM_SR_ERR_M);
+
+  /* Erase and program option in one go. */
+  adiv5_ap_mem_write (ap, STM32Lx_NVM_PECR(nvm),
+                      (is_stm32l1 ? 0 : STM32Lx_NVM_PECR_DATA)
+                      | STM32Lx_NVM_PECR_FIX);
+  if (cb == 4)
+    adiv5_ap_mem_write (ap, address, value);
+  else if (cb == 2)
+    adiv5_ap_mem_write_halfword (ap, address, value);
+  else
+    return false;
+
+  uint32_t sr;
+  do {
+    sr = adiv5_ap_mem_read (ap, STM32Lx_NVM_SR(nvm));
+  } while (sr & STM32Lx_NVM_SR_BSY);
+
+  return !(sr & STM32Lx_NVM_SR_ERR_M);
+}
+
+static bool stm32lx_cmd_stubs (target* t,
+                               int argc, char** argv)
+{
+  if (argc == 1) {
+    gdb_out ("usage: mon stubs [enable/disable]\n");
+  }
+  else if (argc == 2) {
+    size_t cb = strlen (argv[1]);
+    if (!strncasecmp (argv[1], "enable", cb))
+      inhibit_stubs = 0;
+    if (!strncasecmp (argv[1], "disable", cb))
+      inhibit_stubs = 1;
+  }
+  gdb_outf ("stubs: %sabled\n", inhibit_stubs ? "dis" : "en");
+
+  return true;
+}
+
+#if 0
```

Comment:
I can remove it if you prefer.  The hitch with this function is that it doesn't work because the reset of the emulator code gets confused by the target behavior when we perform the OBL_LAUNCH.  I would hope that we could leave this in a work toward solving the option byte loader issue.  However, if you prefer I'll remove it and leave it in my independent tree.